XenAdmin/Dialogs/NewTagDialog.cs

```diff
@@ -302,117 +237,39 @@ private void addButton_Click(object sender, EventArgs e)
-                action.RunAsync();
-            }
-        }
-
         private void textBox1_TextChanged(object sender, EventArgs e)
         {
             addButton.Enabled = (this.textBox1.Text.Trim() != string.Empty);
         }
 
-        private void tagsListView_SelectedIndexChanged(object sender, EventArgs e)
-        {
-            setButtonEnablement();
-        }
-
-        private void setButtonEnablement()
-        {
-            renameButton.Enabled = tagsListView.SelectedItems.Count == 1;
-            deleteButton.Enabled = tagsListView.SelectedItems.Count == 1;
-        }
-
-        private void tagsListView_MouseClick(object sender, MouseEventArgs e)
-        {
-            TagsListViewItem item = (TagsListViewItem)tagsListView.GetItemAt(e.X, e.Y);
-
-            if (item != null)
-            {
-                Rectangle iconRect = tagsListView.GetItemRect(item.Index, ItemBoundsPortion.Icon);
-                Rectangle checkRect = new Rectangle(0, iconRect.Top, iconRect.Left, iconRect.Height);
-
-                // if there's a selection and the clicked item is in the selection, then toggle the
-                // entire selection
-
-                if (checkRect.Contains(e.Location))
-                {
-                    if (tagsListView.SelectedItems.Count > 0 && tagsListView.SelectedItems.Contains(item))
-                    {
-                        ToggleItems(tagsListView.SelectedItems);
-                    }
-                    else
-                    {
-                        item.Toggle();
-                    }
-                }
-            }
-        }
-
-        private void tagsListView_KeyDown(object sender, KeyEventArgs e)
+        private void tagsDataGrid_KeyDown(object sender, KeyEventArgs e)
         {
             if (e.KeyCode == Keys.Space || e.KeyCode == Keys.Enter)
             {
-                ToggleItems(tagsListView.SelectedItems);
+                ToggleItems(tagsDataGrid.SelectedRows);
+                e.Handled = true;
             }
         }
 
         private void ToggleItems(System.Collections.IList items)
         {
-            if(items.Count < 1)
+            if (items.Count < 1)
                 return;
 
-            CheckState firstCheckState = ((TagsListViewItem)items[0]).Checked;
-            foreach (TagsListViewItem item in items)
+            CheckState firstCheckState = ((TagsDataGridViewRow) items[0]).Checked;
+            foreach (TagsDataGridViewRow item in items)
             {
                 item.Toggle(firstCheckState);
             }
         }
 
-        private class TagsListViewItem : ListViewItem
+        public class TagsDataGridViewRow : DataGridViewRow
         {
-            public TagsListViewItem(string text)
-                : base(text)
+            public TagsDataGridViewRow()
             {
-                Checked = CheckState.Unchecked;
+                var cellEnabled = new DataGridViewCheckBoxCell { Value = CheckState.Unchecked };

```

Comment:
Please also set ThreeState = true on the CheckBoxCell (maybe on the column as well), otherwise the Indeterminate state does not appear correctly. To test this, set, for example, tag Foo on vm1, tag Bar on vm2, then select both VMs and right click Edit tags. Foo and Bar currently appear Checked instead of Indeterminate.